kratos/utilities/auxilary_global_master_slave_relation.h

```diff
@@ -0,0 +1,275 @@
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(AUXILARY_GLOBAL_MASTER_SLAVE_RELATION)
+#define AUXILARY_GLOBAL_MASTER_SLAVE_RELATION
+// System includes
+#include <vector>
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include "includes/lock_object.h"
+
+
+namespace Kratos
+{
+
+/**
+ * @class AuxilaryGlobalMasterSlaveRelation
+ * @ingroup KratosCore
+ * @brief This class stores the information regarding the AuxilaryGlobalMasterSlaveRelation equation.
+ *         Naming convenction is defined like this. (each object of this class will store one equation in the given form
+ *
+ *   SlaveEquationId = w_1*MasterEquationId_1 + w_2*MasterEquationId_2 + ..... + w_n*MasterEquationId_n
+ *
+ *   This stores the condensed form of the MasterSlaveConstraint objects into one object. if only one relation for a slave is added as
+ *   MasterSlaveConstraint then there will only be one entry for master for its corresponding AuxilaryGlobalMasterSlaveRelation.
+ *   Currently this class is designed to hold only one equation. There is only one unique object of this class for each slave. 
+ *
+ *   Future plan is to also make it possible to work with matrices (T) and vectors (for slave and master equation ids and constants)
+ *
+ *
+ *  IMPORTANT : This is not seen by the user. This is a helper data structure which is exists only in the builder and solver. 
+ *
+ * @author Aditya Ghantasala
+ */
+class AuxilaryGlobalMasterSlaveRelation : public IndexedObject
+{
+  public:
+    typedef IndexedObject BaseType;
+    typedef BaseType::IndexType IndexType;
+    typedef Matrix MatrixType;
+    typedef Vector VectorType;
+    typedef std::vector<IndexType> EquationIdVectorType;
+    KRATOS_CLASS_POINTER_DEFINITION(AuxilaryGlobalMasterSlaveRelation);
+
+    /**
+     * @brief Constructor of the class
+     * @param SlaveEquationId the slave equation id for which this class is being constructed.
+     */
+    AuxilaryGlobalMasterSlaveRelation(IndexType SlaveEquationId = 0) : IndexedObject(SlaveEquationId), mConstant(0.0)
+    {
+    }
+
+    /**
+     * @brief Function to get the slave equation Id corresponding to this constraint.
+     * @param Constant the value of the constant to be assigned.
+     */
+    IndexType SlaveEquationId() const { return this->Id(); }
+
+    /**
+     * @brief Function to set the lefthand side of the constraint (the slave dof value)
+     * @param LhsValue the value of the lhs (the slave dof value)
+     */
+    void SetLHSValue(const double LhsValue)
+    {
+        mLockObject.SetLock();
```

Comment:
There is no non-deterministic value possible here as the slave (DOF) can have only one value. So here every thread is trying to set the same value here. I put the locks here to be safe from runtime termination or memory errors. Also about calling this function in in Parallel region : I was calling this in the function in here https://github.com/KratosMultiphysics/Kratos/blob/9b2df88f88d23c7df93c12de665ea25bb3c808c0/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h#L693

So I could not avoid calling it inside parallel region. So the locks.

Reply:
In that case the `it` in `(it->second)->SetLHSValue(0.0);` is not a shared variable so it is safe to use without the locks. In case `it` would be a shared variable there are two possibilities: 1) some developer tries to initialize the lhs to the same value using multiple threads, which should not be done for performance reasons, and 2) the developer tries to write different values from different threads, which will be non-deterministic and most definitely a bug.